src/crypto.c

```diff
@@ -139,12 +146,19 @@ size_t kv_decrypt(struct kv_crypto_st *kvmgc, decrypt_callback cb, void *userdat
         err = crypto_skcipher_decrypt(kvmgc->req);
         if (err) {
             prerr("Decryption failed\n");
+            memcpy(kvmgc->iv, iv_orig, sizeof(kvmgc->iv));
+            memcpy(kvmgc->kv_data.buf, data_orig, buflen);
+            kfree(data_orig);
```

Comment:
can we factor this out into a separate `static` function that will serve as a helper for the cleanup? In addition, this code that was duplicated at several places.